### PR TITLE
Gardening for `@dynamicMemberLookup`.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Unify code style of `@dynamicMemberLookup` and `@dynamicCallable` implementations.
  - Use consistent variable names, diagnostic names/messages, doc comments, etc.
- Move `@dynamicMemberLookup` test to `test/attr` directory.

New consistent naming:
- `invalid_dynamic_callable_type`, `invalid_dynamic_member_lookup_type`
- `DynamicCallableCache`, `DynamicMemberLookupCache`
- `isValidDynamicCallableMethod`, `isValidDynamicMemberLookupSubscript`

This PR is almost NFC. The only user-facing change is a new error message for `invalid_dynamic_member_lookup_type`.

### CODE REVIEW

**Comment by dan-zheng** [comment]
cc @lattner: could you please review?

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -1054,9 +1054,10 @@ ERROR(missing_dynamic_callable_kwargs_method,none,
       "@dynamicCallable type %0 cannot be applied with keyword arguments; "
       "missing 'dynamicCall(withKeywordArguments:)' method", (Type))
 
-ERROR(type_invalid_dml,none,
+ERROR(invalid_dynamic_member_lookup_type,none,
       "@dynamicMemberLookup attribute requires %0 to have a "
-      "'subscript(dynamicMember:)' member with a string index", (Type))
+      "'subscript(dynamicMember:)' method with an "
+      "'ExpressibleByStringLiteral' argument", (Type))
```
Comment (dan-zheng) [inline include/swift/AST/DiagnosticsSema.def]: There are a few names for "argument to subscript method": "argument"/"parameter" and "index".
I think "argument" is a bit easier to understand (unmistakable meaning).

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -3053,17 +3053,17 @@ getArgumentLabels(ConstraintSystem &cs, ConstraintLocatorBuilder locator) {
 /// conformances, but this is fine because it doesn't get invoked in the normal
 /// name lookup path (only when lookup is about to fail).
 static bool hasDynamicMemberLookupAttribute(CanType ty,
```
Comment (dan-zheng) [inline lib/Sema/CSSimplify.cpp]: Deduping this method with `getDynamicCallableMethods` (which also defines the same "type traversal" logic) would be nice. However, the two methods use different cache types (`llvm::DenseMap<CanType, bool>` vs `llvm::DenseMap<CanType, DynamicCallableMethods>`), which makes generalization difficult. 🙁

Any generalization ideas appreciated.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (xwu) [reply include/swift/AST/DiagnosticsSema.def]: Here, I _think_ the right term would be “parameter,” would it not? As in, one passes arguments to methods that have parameters.

Comment (dan-zheng) [reply include/swift/AST/DiagnosticsSema.def]: Yes, I agree. 😃
Renamed "argument" to "parameter" in https://github.com/apple/swift/pull/20498/commits/18a8ea911e647867c392829bdda94ca62a48855e.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by lattner** [comment]
This all LGTM, but please fix the attr/attr_dynamic_member_lookup.swift failure.

**Review by lattner — APPROVED** [review]

**Comment by dan-zheng** [comment]
Sorry, tests failed because I updated the diagnostic message without updating tests.
@swift-ci Please smoke test

**Pull request merged**